### ISSUE TITLE
newbie question and a few bugs found

### ISSUE THREAD

**Issue by jcyuan** [mono/CppSharp#1771]
I'm using the default `Driver` to generate cs files:

![image](https://github.com/mono/CppSharp/assets/850854/75bb4c1d-1a42-4d84-a2bc-c44990452011)
this is the code to do the setup

![image](https://github.com/mono/CppSharp/assets/850854/2bf098d6-794e-4c0d-8843-b7b44d975814)
and this is what i get...

btw another question is, why it generates `*-symbols.cpp` files for me? i don't need them, so how to avoid these `*.cpp` files being generated?

so basically 3 questions:
1, duplicated namespace (how to avoid generating the namespace of the added moudle)
2, avoid generating `*.cpp` files
![image](https://github.com/mono/CppSharp/assets/850854/bb79d94c-094c-4a55-b7de-7e26b55a3bef)
3, how to fix this error? (why it's not `new Color(a, b, c)`)? or, how to add a new constructor for a class?

sorry for these questions, i have tried to find on Google or in the Docs but no luck.
thanks so much.

**Comment by jcyuan** [comment]
answer myself:

1, just set `newMoudle.OutputNamespace = string.Empty`
2, didn't find the snippet where handled it in the source code
3, didn't find the snippet where handled it in the source code, i have just simpled ignored them with an ugly way, because these global fields are just for handy purpose, so can be ignored.
```
// ugly but works, maybe there is more elegant way?
var ignoreList = new []
        {
            "Black",
            "White",
            "Red",
            "Green",
            "Blue"
        };

        foreach (var u in ctx.TranslationUnits)
        {
            foreach (var ns in u.Namespaces)
            {
                foreach (var v in ns.Variables)
                {
                    if (Array.IndexOf(ignoreList, v.Name) >= 0)
                        v.Ignore = true;
                }
            }
        }
```

in the `UsersManu.md` it says global scope functions are handled in the `FunctionToXXXXMethodPass`, but i didn't find which pass handles for variables.

**Comment by jcyuan** [comment]
```
struct Point {
float length() const {
    return Point::Length(x, y);
  }
  static float Length(float x, float y) {
    return sqrt(x * x + y * y);
  }
}
```

generate this will cause error too.

**Comment by jcyuan** [comment]
```
enum class Backend {
  /**
   * Mock is a backend that does not draw anything. It is used for unit tests and to measure CPU
   * overhead.
   */
  MOCK,
  OPENGL,
  METAL,
  VULKAN,
};
```

the generated comments are wrong:

```
public enum Backend
    {
        /// <summary>
        /// <para>Mock is a backend that does not draw anything. It is used for unit tests and to measure CPU</para>
        /// <para>overhead.</para>
        /// </summary>
        MOCK = 0,
        /// <summary>
        /// <para>Mock is a backend that does not draw anything. It is used for unit tests and to measure CPU</para>
        /// <para>overhead.</para>
        /// </summary>
        OPENGL = 1,
        /// <summary>
        /// <para>Mock is a backend that does not draw anything. It is used for unit tests and to measure CPU</para>
        /// <para>overhead.</para>
        /// </summary>
        METAL = 2,
        /// <summary>
        /// <para>Mock is a backend that does not draw anything. It is used for unit tests and to measure CPU</para>
        /// <para>overhead.</para>
        /// </summary>
        VULKAN = 3
    }
```

**Comment by jcyuan** [comment]
thanks for youre reply @tritao 
later if i have time i'll try to PR,
here another i just found:

```
public virtual void GenerateUsings()
        {
//            .......
            foreach (var library in internalsVisibleTo)
                WriteLine($"[assembly:InternalsVisibleTo(\"{library}\")]");    
```
`InternalsVisibleTo` needs a feature that consider for public key if the current library has one.

**Comment by jcyuan** [comment]
https://github.com/mono/CppSharp/issues/968

i'm encountering this, is there a easy way to fix it?

**Comment by jcyuan** [comment]
well, everything is fine but stuck on `std::shared_ptr`.... a lot of key methods are ignored...
don't know what to do now.....

https://github.com/mono/CppSharp/issues/1738

not sure how did @frankenherm do it.

**Comment by jcyuan** [comment]
![image](https://github.com/mono/CppSharp/assets/850854/34d10984-3373-4590-ac05-3d4ecb732ed3)

hi @tritao, @ddobrev for such a complex type, what is the right way to write a `TypeMap` for it?
i have to fully figure out how to customize type first before pushing any merge request.

if just a simple type, use `[TypeMap("typestr", GeneratorKind = GeneratorKind.CSharp)]` would be enough, 
but with such an complex one, how to define?
`[TypeMap("::std::_Ptr_base", GeneratorKind = GeneratorKind.CSharp)]`, obviously this is not enough, so how to define `::std::_Ptr_base<???>::element_type*`? because ??? can be any types.

thanks

**Comment by jcyuan** [comment]
> I don't think that's supported atm.
> 
> Check here how we do it for strings: https://github.com/mono/CppSharp/blob/main/src/Generator/Types/Std/Stdlib.CSharp.cs#L302
> 
> This code could be extended to do a regex match over all the typemaps: https://github.com/mono/CppSharp/blob/main/src/Generator/Types/TypeMapDatabase.cs#L63
> 
> Then you could have something like `*` in the type names. For now you can have multiple `[TypeMap(...)]` for multiple types you might need.

thanks so much for the tips, i'll have a look now. 👍

**Comment by jcyuan** [comment]
![image](https://github.com/mono/CppSharp/assets/850854/7178107c-30d4-420b-bb90-fde19374deb1)
![image](https://github.com/mono/CppSharp/assets/850854/6666b8ea-07e2-4798-8929-c0a7d6d5151c)
i have both `Property<T>` and `Point` classes generated successfully, but why `Property<Point>*` is an incomplete type?
i have checked in the source code but didn't find the reason (in `CheckIgnoredDeclsPass`), and even define a `[TypeMap("::pag::Property<::pag::Point>*", GeneratorKind = GeneratorKind.CSharp)]` for it but it still does not work....
really have no idea on how to fix it now, could you please help me on where should i go? @tritao 
thanks for your time.

**Comment by jcyuan** [comment]
> Want to help but it's hard to say based on the info. These messages should be improved to give more information to the user, for now I would recommend you add some breakpoints and use the debugger to help figure out what's going on.

well, thanks so much, actually i did so in the source code already, mainly it's because the source code does not have comments so a little hard to understand some of the parsing details, although the main parsing flow is easy to understand.

**Comment by jcyuan** [comment]
> > > Want to help but it's hard to say based on the info. These messages should be improved to give more information to the user, for now I would recommend you add some breakpoints and use the debugger to help figure out what's going on.
> > 
> > 
> > well, thanks so much, actually i did so in the source code already, mainly it's because the source code does not have comments so a little hard to understand some of the parsing details, although the main parsing flow is easy to understand.
> 
> One technique you might try:
> 
> https://github.com/mono/CppSharp/blob/main/src/AST/Declaration.cs#L157
> 
> Override the base `set` with something like:
> 
> ```cs
>             set => { if (Name == "::pag::Property<::pag::Point>*") System.Diagnostics.Debugger.Break(); }
> ```
> 
> Then your debugger will break when your AST node is being ignored. Just make sure the `Name` is right, but the overall technique is how I debug these kind of issues.

okay 👍, thanks man
